### PR TITLE
Idefics: fix position ids

### LINKED ISSUE #33852: IDEFICS cannot work with past_key_values when I'm using prefix tuning from peft library.
### System Info

- `transformers` version: 4.44.2
- Platform: Linux-4.15.0-76-generic-x86_64-with-glibc2.27
- Python version: 3.12.4
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.34.2
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: DEEPSPEED
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 4
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - deepspeed_config: {'gradient_accumulation_steps': 16, 'gradient_clipping': 1.0, 'offload_optimizer_device': 'cpu', 'offload_param_device': 'cpu', 'zero3_init_flag': False, 'zero3_save_16bit_model': False, 'zero_stage': 3}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- PyTorch version (GPU?): 2.3.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA GeForce RTX 3090

### Who can help?

@ArthurZucker @amyeroberts, @qubvel

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import peft
from transformers import IdeficsForVisionText2Text, AutoProcessor
import sys
import torch

sys.path.insert(0, "..")
import config  # my own file

device = torch.device("cuda:2")
model = IdeficsForVisionText2Text.from_pretrained(
    config.idefics_9b_path, torch_dtype=torch.float16
).to(device)
processor = AutoProcessor.from_pretrained(
    config.idefics_9b_path, torch_dtype=torch.float16
)
model = peft.get_peft_model(
    model,
    peft.PrefixTuningConfig(
        peft_type="PREFIX_TUNING",
        task_type="CAUSAL_LM",
        num_virtual_tokens=1,
        token_dim=4096,
        num_transformer_submodules=1,
        num_attention_heads=32,
        num_layers=32,
        encoder_hidden_size=768,
    ),
)
inputs = processor(["hello"]).to(device)
model(**inputs)
```

### Expected behavior

When I ran above code snippet, I got:
```
{
	"name": "RuntimeError",
	"message": "The size of tensor a (2) must match the size of tensor b (3) at non-singleton dimension 2",
	"stack": "---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[3], line 2
      1 inputs = inputs.to(device)
----> 2 model(**inputs)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1532, in Module._wrapped_call_impl(self, *args, **kwargs)
   1530     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1531 else:
-> 1532     return self._call_impl(*args, **kwargs)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1541, in Module._call_impl(self, *args, **kwargs)
   1536 # If we don't have any hooks, we want to skip the rest of the logic in
   1537 # this function, and just call forward.
   1538 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1539         or _global_backward_pre_hooks or _global_backward_hooks
   1540         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1541     return forward_call(*args, **kwargs)
   1543 try:
   1544     result = None

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/peft/peft_model.py:1680, in PeftModelForCausalLM.forward(self, input_ids, attention_mask, inputs_embeds, labels, output_attentions, output_hidden_states, return_dict, task_ids, **kwargs)
   1677 if peft_config.peft_type == PeftType.PREFIX_TUNING:
   1678     # overwrite past_kv in kwargs
   1679     kwargs[\"past_key_values\"] = self.get_prompt(batch_size)
-> 1680     return self.base_model(input_ids=input_ids, inputs_embeds=inputs_embeds, **kwargs)
   1681 else:
   1682     if inputs_embeds is None:

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1532, in Module._wrapped_call_impl(self, *args, **kwargs)
   1530     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1531 else:
-> 1532     return self._call_impl(*args, **kwargs)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1541, in Module._call_impl(self, *args, **kwargs)
   1536 # If we don't have any hooks, we want to skip the rest of the logic in
   1537 # this function, and just call forward.
   1538 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1539         or _global_backward_pre_hooks or _global_backward_hooks
   1540         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1541     return forward_call(*args, **kwargs)
   1543 try:
   1544     result = None

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/transformers/models/idefics/modeling_idefics.py:1496, in IdeficsForVisionText2Text.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, pixel_values, image_encoder_embeddings, perceiver_embeddings, image_attention_mask, labels, use_cache, output_attentions, output_hidden_states, interpolate_pos_encoding, return_dict)
   1493 return_dict = return_dict if return_dict is not None else self.config.use_return_dict
   1495 # decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)
-> 1496 outputs = self.model(
   1497     input_ids=input_ids,
   1498     attention_mask=attention_mask,
   1499     position_ids=position_ids,
   1500     past_key_values=past_key_values,
   1501     inputs_embeds=inputs_embeds,
   1502     pixel_values=pixel_values,
   1503     image_encoder_embeddings=image_encoder_embeddings,
   1504     perceiver_embeddings=perceiver_embeddings,
   1505     image_attention_mask=image_attention_mask,
   1506     use_cache=use_cache,
   1507     output_attentions=output_attentions,
   1508     output_hidden_states=output_hidden_states,
   1509     interpolate_pos_encoding=interpolate_pos_encoding,
   1510     return_dict=return_dict,
   1511 )
   1513 hidden_states = outputs[0]
   1514 logits = self.lm_head(hidden_states)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1532, in Module._wrapped_call_impl(self, *args, **kwargs)
   1530     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1531 else:
-> 1532     return self._call_impl(*args, **kwargs)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1541, in Module._call_impl(self, *args, **kwargs)
   1536 # If we don't have any hooks, we want to skip the rest of the logic in
   1537 # this function, and just call forward.
   1538 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1539         or _global_backward_pre_hooks or _global_backward_hooks
   1540         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1541     return forward_call(*args, **kwargs)
   1543 try:
   1544     result = None

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/transformers/models/idefics/modeling_idefics.py:1329, in IdeficsModel.forward(self, input_ids, attention_mask, position_ids, past_key_values, inputs_embeds, pixel_values, image_encoder_embeddings, perceiver_embeddings, image_attention_mask, use_cache, output_attentions, output_hidden_states, interpolate_pos_encoding, return_dict)
   1312     layer_outputs = self._gradient_checkpointing_func(
   1313         vblock,
   1314         decoder_layer,
   (...)
   1326         self.gated_cross_attn_layers,
   1327     )
   1328 else:
-> 1329     layer_outputs = vblock(
   1330         decoder_layer,
   1331         hidden_states,
   1332         attention_mask=attention_mask,
   1333         position_ids=position_ids,
   1334         past_key_value=past_key_value,
   1335         image_hidden_states=image_hidden_states,
   1336         image_attention_mask=image_attention_mask,
   1337         cross_attention_gate=cross_attention_gate,
   1338         output_attentions=output_attentions,
   1339         use_cache=use_cache,
   1340         layer_idx=idx,
   1341         cross_layer_interval=self.cross_layer_interval,
   1342         gated_cross_attn_layers=self.gated_cross_attn_layers,
   1343     )
   1345 hidden_states = layer_outputs[0]
   1347 if use_cache:

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/transformers/models/idefics/modeling_idefics.py:1293, in IdeficsModel.forward.<locals>.vblock(main_block, hidden_states, attention_mask, position_ids, past_key_value, image_hidden_states, image_attention_mask, cross_attention_gate, output_attentions, use_cache, layer_idx, cross_layer_interval, gated_cross_attn_layers)
   1281     outputs = xblock(
   1282         hidden_states,
   1283         attention_mask=attention_mask,
   (...)
   1289         past_key_value=None,  # not implemented
   1290     )
   1291     hidden_states = outputs[0]
-> 1293 layer_outputs = main_block(
   1294     hidden_states,
   1295     attention_mask=attention_mask,
   1296     position_ids=position_ids,
   1297     past_key_value=past_key_value,
   1298     output_attentions=output_attentions,
   1299     use_cache=use_cache,
   1300 )
   1302 return layer_outputs

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1532, in Module._wrapped_call_impl(self, *args, **kwargs)
   1530     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1531 else:
-> 1532     return self._call_impl(*args, **kwargs)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1541, in Module._call_impl(self, *args, **kwargs)
   1536 # If we don't have any hooks, we want to skip the rest of the logic in
   1537 # this function, and just call forward.
   1538 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1539         or _global_backward_pre_hooks or _global_backward_hooks
   1540         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1541     return forward_call(*args, **kwargs)
   1543 try:
   1544     result = None

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/transformers/models/idefics/modeling_idefics.py:749, in IdeficsDecoderLayer.forward(self, hidden_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache)
    746 hidden_states = self.input_layernorm(hidden_states)
    748 # Self Attention
--> 749 hidden_states, self_attn_weights, present_key_value = self.self_attn(
    750     hidden_states=hidden_states,
    751     attention_mask=attention_mask,
    752     position_ids=position_ids,
    753     past_key_value=past_key_value,
    754     output_attentions=output_attentions,
    755     use_cache=use_cache,
    756 )
    757 hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
    758 hidden_states = residual + hidden_states

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1532, in Module._wrapped_call_impl(self, *args, **kwargs)
   1530     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1531 else:
-> 1532     return self._call_impl(*args, **kwargs)

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/torch/nn/modules/module.py:1541, in Module._call_impl(self, *args, **kwargs)
   1536 # If we don't have any hooks, we want to skip the rest of the logic in
   1537 # this function, and just call forward.
   1538 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1539         or _global_backward_pre_hooks or _global_backward_hooks
   1540         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1541     return forward_call(*args, **kwargs)
   1543 try:
   1544     result = None

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/transformers/models/idefics/modeling_idefics.py:640, in IdeficsAttention.forward(self, hidden_states, key_value_states, attention_mask, position_ids, past_key_value, output_attentions, use_cache)
    638 if not is_cross_attention:
    639     cos, sin = self.rotary_emb(value_states, seq_len=max(kv_seq_len, q_len))
--> 640     query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
    641 # [bsz, nh, t, hd]
    643 if past_key_value is not None:
    644     # reuse k, v, self_attention

File ~/miniconda3/envs/icl/lib/python3.12/site-packages/transformers/models/idefics/modeling_idefics.py:509, in apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim)
    507 cos = cos[position_ids].unsqueeze(unsqueeze_dim)
    508 sin = sin[position_ids].unsqueeze(unsqueeze_dim)
--> 509 q_embed = (q * cos) + (rotate_half(q) * sin)
    510 k_embed = (k * cos) + (rotate_half(k) * sin)
    511 return q_embed, k_embed

RuntimeError: The size of tensor a (2) must match the size of tensor b (3) at non-singleton dimension 2"
}
```

As we can see, due to appearance of `past_key_values`, the key states and the value states will be extent to a larger length, while query states won't. The `cos` here is generated by `position_ids` whose length is same as `attention_mask`.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/33852 by cropping position ids to the length of inputs and cleans up a little bit the way inputs are prepared for generation

TF equivalence tests are failing for me locally and also failing on `main`, so those are not related to this PR

### CODE REVIEW

File: src/transformers/models/idefics/modeling_idefics.py [hunk 1]
```diff
@@ -1212,11 +1167,9 @@ def forward(
             # create position_ids on the fly for batch generation
             position_ids = attention_mask.long().cumsum(-1) - 1
             position_ids.masked_fill_(attention_mask == 0, 1)
+            position_ids = position_ids[:, -seq_length:]
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:1170]: The attention mask is of shape `seq_length_with_past`, so that makes sense!

**Review by ArthurZucker — APPROVED** [review]
Thanks, there might be a test worth adding for this no? (we break that often for peft 😄 )

**Comment by Kami-chanw** [comment]
When this branch can be merged?

**Comment by zucchini-nlp** [comment]
Will take a look at your comment and merge after seeing what is wrong with labels. TBH. labels should not be affected by position ids, but I'll dig into

**Comment by zucchini-nlp** [comment]
@ArthurZucker can you take one more look pls? I added fix for `labels`. Seems one of the very first VLMs introduced it and all other simply copied from that one. I don't see any attn mask applied in LLMs and I don't think we should be doing it under the hood, without the user knowing. So I deprecated that for 2 minor releases

**Comment by BenjaminBossan** [comment]
> Thanks, there might be a test worth adding for this no? (we break that often for peft 😄 )

Maybe PEFT would be a better place to add this test. If so, LMK and I can work on something (probably based on [this code](https://github.com/huggingface/transformers/issues/33852#issuecomment-2389184690)). Since that would only test a single architecture, it wouldn't be super useful IMO but still better than nothing.

File: src/transformers/models/idefics2/modeling_idefics2.py [hunk 2]
```diff
@@ -1626,7 +1626,14 @@ def forward(
...
             if attention_mask is not None:
-                shift_attention_mask = attention_mask[..., 1:].to(logits.device)
+                # we use the input attention mask to shift the logits and labels, because it is 2D.
+                # we also crop attn mask in case it is longer, which happens in PrefixTuning with peft
+                shift_attention_mask = attention_mask[:, -(logits.shape[1] - 1) :].to(logits.device)
+                logger.warning_once(
+                    "The final logits were masked using attention mask before calculating the loss. "
+                    "This behavior will be removed in v4.48, you should be masking the `labels` with `-100` "
+                    "in data collator before training starts."
+                )
```
Comment (ArthurZucker) [inline src/transformers/models/idefics2/modeling_idefics2.py:1636]: that's the only thing I am curious about, we are introducing something new that we know we are gonna deprecate. Not sure this makes a lot of sense to me! Would either not add this, or just not progagate to other models than idefics

**Review by ArthurZucker — APPROVED** [review]
Just missing info on the shift, we broke this recently (I think we never had this) would be in favor of doing less work for the users, but properly documenting !

File: src/transformers/models/idefics2/modeling_idefics2.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/idefics2/modeling_idefics2.py:1636]: Yes, I don't know how this got introduced as none of our llms mask out labels. We can also not deprecate and stop propagating, yes. I'll take care of it when new models are added

So, in that case we remove deprecation and leave the fix for PEFT tuning

Comment (ArthurZucker) [reply src/transformers/models/idefics2/modeling_idefics2.py:1636]: > So, in that case we remove deprecation and leave the fix for PEFT tuning

exactly! We already get bashed enough for the platora of warning we produce, let's not add one more! 😉

Comment (BenjaminBossan) [reply src/transformers/models/idefics2/modeling_idefics2.py]: Thanks for addressing the issue in this PR.

> So, in that case we remove deprecation and leave the fix for PEFT tuning

What fix are we talking about here?

Comment (zucchini-nlp) [reply src/transformers/models/idefics2/modeling_idefics2.py]: oh, that's more transformer-side fix. We are talking about attention mask, which gets extra virtual tokens with PEFT so we have to crop them off to match the shape of `labels` for CELoss

Comment (BenjaminBossan) [reply src/transformers/models/idefics2/modeling_idefics2.py]: Got it, thanks, I thought I might have to fix something in PEFT.

**Pull request merged**